wrappers/python/simtk/openmm/app/simulation.py

```diff
@@ -198,25 +201,57 @@ def _simulate(self, endStep=None, endTime=None):
             self.integrator.step(stepsToGo)
             self.currentStep += nextSteps
             if anyReport:
-                getPositions = False
-                getVelocities = False
-                getForces = False
-                getEnergy = False
-                for reporter, next in zip(self.reporters, nextReport):
-                    if next[0] == nextSteps:
-                        if next[1]:
-                            getPositions = True
-                        if next[2]:
-                            getVelocities = True
-                        if next[3]:
-                            getForces = True
-                        if next[4]:
-                            getEnergy = True
-                state = self.context.getState(getPositions=getPositions, getVelocities=getVelocities, getForces=getForces,
-                                              getEnergy=getEnergy, getParameters=True, enforcePeriodicBox=self._usesPBC)
-                for reporter, next in zip(self.reporters, nextReport):
-                    if next[0] == nextSteps:
-                        reporter.report(self, state)
+                # One or more reporters are ready to generate reports.  Organize them into three
+                # groups: ones that want wrapped positions, ones that want unwrapped positions,
+                # and ones that don't care about positions.
+                
+                wrapped = []
+                unwrapped = []
+                either = []
+                for reporter, report in zip(self.reporters, nextReport):
+                    if report[0] == nextSteps:
+                        if len(report) > 5:
+                            wantWrap = report[5]
+                            if wantWrap is None:
+                                wantWrap = self._usesPBC
+                        else:
+                            wantWrap = self._usesPBC
+                        if not report[1]:
+                            either.append((reporter, report))
+                        elif wantWrap:
+                            wrapped.append((reporter, report))
+                        else:
+                            unwrapped.append((reporter, report))
+                if len(wrapped) > len(unwrapped):
+                    wrapped += either
+                else:
+                    unwrapped += either
```

Comment:
From an implementation perspective it seems redundant to recalculate the wrapped/unwrapped reporter lists at each time step, no? I would just move it out of the loop.

Reply:
Oh right, I see.